### PR TITLE
[WIP] Support Continual Pretraining Multi Dataset using Streaming

### PR DESCRIPTION
**Pull request by mostafaelhoushi**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* Added a training script and recipe for continual pretraining. It enables streaming through large datasets from Hugging Face without having to download them, or existing directory with .jsonl files.

I had to make some workarounds to the multidata approach implemented in #1929 . I am not sure if this is because I didn't understand the codebase or if some classes related to data loading and collating need to be refactored.
Cc @andrewkho, @RdoubleA, @ebsmothers 

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: recipes/full_continual_distributed_multi_dataset.py [hunk 1]
```diff
@@ -0,0 +1,948 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (mostafaelhoushi) [inline recipes/full_continual_distributed_multi_dataset.py:1]: All of the contents of this file except the `_setup_data()` function was copied from [recipes/full_finetune_distributed.py](https://github.com/pytorch/torchtune/blob/d6fbab8d02f37555d938e574c1f78e48a6a13ce3/recipes/full_finetune_distributed.py)

Here is the diff comparing the 2 files:
https://www.diffchecker.com/CjtUG31A/

File: recipes/full_continual_distributed_multi_dataset.py [hunk 2]
```diff
@@ -0,0 +1,948 @@
...
+                training.load_from_full_optimizer_state_dict(
+                    optimizer,
+                    opt_state_dict,
+                    self._device,
+                )
+
+            utils.log_rank_zero(log, "Optimizer is initialized.")
+            return optimizer
+
+    def _setup_data(
```
Comment (mostafaelhoushi) [inline recipes/full_continual_distributed_multi_dataset.py:624]: This `_setup_data(..)` function was copied from [recipes/lora_finetune_distributed_multi_dataset.py](https://github.com/pytorch/torchtune/blob/d6fbab8d02f37555d938e574c1f78e48a6a13ce3/recipes/lora_finetune_distributed_multi_dataset.py#L569C1-L662C1) and modified.

You can view the modifications in this online Diff:  https://www.diffchecker.com/Cg4AiMni/

**Review by mostafaelhoushi — COMMENTED** [review]
Added some explanation comments
